cranelift/codegen/src/isa/s390x/lower.rs

```diff
@@ -0,0 +1,2807 @@
+    if let Some(insn) = input_matches_insn(ctx, input, Opcode::Sextend) {
+        if ty_bits(ctx.input_ty(insn, 0)) == 32 {
+            let reg = put_input_in_reg(ctx, InsnInput { insn, input: 0 }, NarrowValueMode::None);
+            return Some(reg);
+        }
+    }
+    None
+}
+
+fn input_matches_uext32_reg<C: LowerCtx<I = Inst>>(ctx: &mut C, input: InsnInput) -> Option<Reg> {
+    if let Some(insn) = input_matches_insn(ctx, input, Opcode::Uextend) {
+        if ty_bits(ctx.input_ty(insn, 0)) == 32 {
+            let reg = put_input_in_reg(ctx, InsnInput { insn, input: 0 }, NarrowValueMode::None);
+            return Some(reg);
+        }
+    }
+    None
+}
+
+fn input_matches_uext16_mem<C: LowerCtx<I = Inst>>(
+    ctx: &mut C,
+    input: InsnInput,
+) -> Option<MemArg> {
+    if ty_bits(ctx.input_ty(input.insn, input.input)) == 16 {
+        return input_matches_load_insn(ctx, input, Opcode::Load);
+    }
+    if ty_bits(ctx.input_ty(input.insn, input.input)) >= 32 {
+        return input_matches_load_insn(ctx, input, Opcode::Uload16);
+    }
+    None
+}
+
+fn input_matches_uext32_mem<C: LowerCtx<I = Inst>>(
+    ctx: &mut C,
+    input: InsnInput,
+) -> Option<MemArg> {
+    if ty_bits(ctx.input_ty(input.insn, input.input)) > 32 {
+        return input_matches_load_insn(ctx, input, Opcode::Uload32);
+    }
+    None
+}
+
+//============================================================================
+// Lowering: force instruction input into a register
+
+/// How to handle narrow values loaded into registers; see note on `narrow_mode`
+/// parameter to `put_input_in_*` below.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+enum NarrowValueMode {
+    None,
+    /// Zero-extend to 32 bits if original is < 32 bits.
+    ZeroExtend32,
+    /// Sign-extend to 32 bits if original is < 32 bits.
+    SignExtend32,
+    /// Zero-extend to 64 bits if original is < 64 bits.
+    ZeroExtend64,
+    /// Sign-extend to 64 bits if original is < 64 bits.
+    SignExtend64,
+}
+
+fn extend_memory_to_reg<C: LowerCtx<I = Inst>>(
+    ctx: &mut C,
+    mem: MemArg,
+    from_ty: Type,
+    to_ty: Type,
+    signed: bool,
+) -> Reg {
+    let rd = ctx.alloc_tmp(to_ty).only_reg().unwrap();
+    ctx.emit(match (signed, ty_bits(to_ty), ty_bits(from_ty)) {
+        (false, 32, 1) => Inst::Load32ZExt8 { rd, mem },
+        (false, 32, 8) => Inst::Load32ZExt8 { rd, mem },
+        (false, 32, 16) => Inst::Load32ZExt16 { rd, mem },
+        (true, 32, 1) => Inst::Load32ZExt8 { rd, mem },
+        (true, 32, 8) => Inst::Load32SExt8 { rd, mem },
+        (true, 32, 16) => Inst::Load32SExt16 { rd, mem },
+        (false, 64, 1) => Inst::Load64ZExt8 { rd, mem },
+        (false, 64, 8) => Inst::Load64ZExt8 { rd, mem },
+        (false, 64, 16) => Inst::Load64ZExt16 { rd, mem },
+        (false, 64, 32) => Inst::Load64ZExt32 { rd, mem },
+        (true, 64, 1) => Inst::Load64ZExt8 { rd, mem },
```

Comment:
I would assume that `Load64SExt8` is wrong here too. The loaded bool is likely stored as a zero extended byte, not sign extended byte and as such it would need to be a load + arithmetic shift left of 63 bits I think.

Reply:
This is also moot.  There actually are no 1-bit memory access instructions in cranelift, so this can never happen.  I've removed all the 1-bit cases here.